### PR TITLE
Focus item in campaign

### PR DESCRIPTION
**Pull request by renjith341**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | No
| New feature? | Yes
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/189
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/3551
| BC breaks? | No
| Deprecations? | No

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Now you could add a focus item as an action in campaign builder. Focus item is a good feature by which we can engage user in our website, but sometimes we may show these to undesired leads. Using focus item inside campaign helps the Mautic user to target focus item for leads from a specific campaign.

#### Steps to test this PR:
1. Create a new campaign with a contact source.
2. Select "Action" 'Show focusitem'.
3. Fill the focus item popup form.
4. Put Mautic page tracker and focus item script in your website
5. Trigger the campaign and test the focus item.

### CODE REVIEW

File: app/bundles/CoreBundle/Assets/js/13.focus.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        return;
+    }
+
+    var url = options.windowUrl;
+    if (url) {
+        var editFocusKey = '/focus/edit/focusId';
+        var previewFocusKey = '/focus/preview/focusId';
+        if (url.indexOf(editFocusKey) > -1 ||
+            url.indexOf(previewFocusKey) > -1) {
+            console.log(url.replace('focusId', mQuery('#campaignevent_properties_focus').val()));
```
Comment (escopecz) [inline app/bundles/CoreBundle/Assets/js/13.focus.js]: This is probably a forgotten debug code

File: app/migrations/Version20170606094013.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('CREATE TABLE '.$this->prefix.'focus_campaign (id INT AUTO_INCREMENT NOT NULL, campaign_id INT DEFAULT NULL, focus_id INT DEFAULT NULL, lead_id INT DEFAULT NULL, INDEX IDX_963C8FF2F639F774 (campaign_id), INDEX IDX_963C8FF251804B42 (focus_id), INDEX IDX_963C8FF255458D (lead_id), PRIMARY KEY(id)) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB');
+        $this->addSql('ALTER TABLE '.$this->prefix.'focus_campaign ADD CONSTRAINT FK_963C8FF2F639F774 FOREIGN KEY (campaign_id) REFERENCES campaigns (id) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE '.$this->prefix.'focus_campaign ADD CONSTRAINT FK_963C8FF251804B42 FOREIGN KEY (focus_id) REFERENCES focus (id) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE '.$this->prefix.'focus_campaign ADD CONSTRAINT FK_963C8FF255458D FOREIGN KEY (lead_id) REFERENCES leads (id) ON DELETE CASCADE');
```
Comment (escopecz) [inline app/migrations/Version20170606094013.php]: These FK names won't work. See the previous migration as an example: https://github.com/mautic/mautic/blob/staging/app/migrations/Version20170506144649.php#L51-L52

Another question is whether you need to create this table. Isn't the `campaing_lead_event_log` sufficient?

**Comment by renjith341** [comment]
@escopecz I created new table 'focus_campaign' since this feature won't be used frequently. During couple of events we need direct relationship between focus, campaign & leads, so I thought it would be better to have separate table other than having a null-able foreign key in campaign events log.

**Comment by kuzmany** [comment]
Should close, already merged https://github.com/mautic/mautic/pull/4666

**Pull request closed without merging**